jurigged/codedb.py

```diff
@@ -0,0 +1,156 @@
+import gc
+import sys
+import time
+import types
+
+from ovld import ovld
+
+##########
+# CodeDB #
+##########
+
+MAX_TIME = 0.1
+
+
+def make_audithook(self):
+    def watch_exec(event, obj):  # pragma: no cover
+        # Note: Python does not trace audit hooks, so normal use will not show
+        # coverage of this function even if it is executed
+        if event == "exec":
+            (code,) = obj
+            self.assimilate(code)
+
+    # The closure is faster than a method on _CodeDB
+    return watch_exec
+
+
+class _CodeDB:
+    def __init__(self):
+        self.codes = {}
+        self.functions = {}
+        self.last_cost = 0
+        self.always_use_cache = False
+
+    def setup(self):
+        self.collect_all()
+        sys.addaudithook(make_audithook(self))
+
+    def collect_all(self):
+        results = []
+        for obj in gc.get_objects():
+            if isinstance(obj, types.FunctionType):
+                results.append((obj, obj.__code__))
+            elif not hasattr(obj, "__getattr__"):
+                if hasattr(obj, "__conform__"):
```

Comment:
What's this about?

Reply:
It's a hook to allow code transforms to update when the original code is changed. For example the `ovld` decorator I'm using to define `conform` in this file would need to do extra work if one of the functions is updated with a different type signature, so it can create an object with the code object in one slot and its `__conform__` method should be called when it changes. Same for a library that JITs a function, etc. I have a bunch of other packages that has to do stuff like this.

Reply:
Oh OK, when I googled `__conform__` I got results about sqlite and a rejected PEP 246 so I was confused.